Blobber Block reward

ssardana96 opened the pull request:
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master BEFORE merging this PR  

Tasks done:
- created a blobber_reward partition
- this blobber_reward partition follows 2 different MPT: 
  + (ongoing_blobber_reward) - this is blobber challenges for current challenge period
  + (active_blobber_reward) - this is blobber challenges for previous challenge period
- change in verifyChallenge to populate these partition
- added blobber block reward functionality
- blobber_block_reward to be triggerred every challenge_period, and not every round
  
Associated PRs (Link as appropriate):
...

ssardana08 committed:
- blobber block reward

ssardana08 committed:
- delegates will get reward from mintReward, so removed code

ssardana08 committed:
- rs.getPartition fixed

ssardana08 committed:
- some neats

ssardana08 committed:
- getblobberlist corrected
- random function corrected
- getRandomSlice handled in unit test

ssardana08 committed:
- blobber_reward itemlist added to partition
- reward blobber to run every challenge_period round
- methods added to PartitionItem and PartitionItemList

ssardana08 committed:
- updated blobber_reward_test

ssardana08 committed:
- updated blobber_block_reward_test

ssardana08 committed:
- updated blobber_block_reward_test

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -330,6 +368,34 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+			return "", common.NewError("verify_challenge",
+				"cannot get ongoing partition: "+err.Error())
+		}
+		err = ongoingList.Migrate(ACTIVE_PASSED_BLOBBERS_KEY, balances)
```
I am not sure what Migrate does but it looks far too complicated. Can't you just do
```go
_, err = balances.InsertTrieNode(ACTIVE_PASSED_BLOBBERS_KEY, ongoingList)
_, err = balances.InsertTrieNode(ONGOING_PASSED_BLOBBERS_KEY, emptyPartition_plus_one_pass_from_this_call)
```

ssardana96 replied:
@Sriep what about the list stored in ongoingList, they also needs to be migrated no?
else the keys will start clashing in future rounds
Partition.Key()

Sriep replied:
The partition name is part of the key for all the individual portions MPT keys.

If you ensure that the block number is part of the overall partition's name then that will be inherited by the key to the individual partitions. So there will be no key cllashing.

ssardana96 replied:
superb. let me do that

Sriep commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -406,17 +472,66 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+				"can't get blobber"+err.Error())
+		}
+
+		ogList, err := getOngoingPassedBlobbersList(balances)
```
Access each MPT key only once. You called `getOngoingPassedBlobbersList` above.

ssardana96 replied:
the reason I'm calling again is because the older one has been migrated and deleted in .Migrate()
this will fetch the latest partition

Sriep replied:
If it has been deleted then you already know that. Use a local variable. 

MPT access takes hundreds of microseconds not nanoseconds. There should be no reason to read or save the same key more than once.

Sriep reviewed (commented):
We still need to decide how to integrate time in seconds with time in rounds.

ssardana08 committed:
- fixed test cases for bbr
- bbr improved

ssardana08 committed:
- some neats

ssardana08 committed:
- benchmark changes for bbr
- changed bbr partition to 10

ssardana08 committed:
- reverting bbr partition to 50

Sriep commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -15,57 +18,71 @@ func (ssc *StorageSmartContract) payBlobberBlockRewards(
 			return common.NewError("blobber_block_rewards_failed",
 				"can't get related stake pool: "+err.Error())
 		}
+
```
you can use [sp.stake](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/stakepool.go#L212) to get the total stake in the stake pool.

ssardana96 replied:
done

Sriep commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -15,57 +18,71 @@ func (ssc *StorageSmartContract) payBlobberBlockRewards(
 
-		capacityReward := float64(conf.BlockReward.BlockReward) * conf.BlockReward.BlobberCapacityWeight * ratio
-		if err := mintReward(qsp, capacityReward, balances); err != nil {
+		if err := mintReward(qsp, reward, balances); err != nil {
```
This will be replaced. We are not going to mint tokens but save to the stake pool rewards using [DistributeRewards](https://github.com/0chain/0chain/blob/358a22285cef5f6782ef78c925295cf266c435d9/code/go/0chain.net/smartcontract/stakepool/stakepool.go#L190). Something like
```go
err = qsp.DistributeRewards(reward, qualifyingBlobberIds[i], stakepool.Blobber, balances)
```
This change requires https://github.com/0chain/0chain/pull/967/

ssardana96 replied:
cool. will make this change once #967 is merged to staging
or should I make this branch merge to #967 ?

Sriep replied:
https://github.com/0chain/0chain/pull/967 has been merged, so we can now make the change; replace `mintReward` with `DistrbuteRewards`.

ssardana08 committed:
- enabling block rewards

ssardana08 committed:
- stake on stakepool called

ssardana08 committed:
- readprice added to partition
- partition movement logic corrected

ssardana08 committed:
- adding totaldata and data read in partition
- saving total data written and read in storage node

ssardana08 committed:
- adding data to partition abot total data and data read
- converting write data to GB

ssardana08 committed:
- gama integration to block reward

ssardana08 committed:
- storage unit test fixed

ssardana08 committed:
- blobber data read logic updated as we would need data read in only last round

ssardana08 committed:
- bbr change calculation added

ssardana08 committed:
- blobber weight added
- removed blobber capacity and blobber usage weight

ssardana08 committed:
- zeta calculation added

ssardana08 committed:
- benchmark corrected

service-0chain commented:
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | bbr  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

ssardana08 committed:
- block reward reverted

peterlimg commented on code/go/0chain.net/miner/protocol_block_main.go:
```diff
@@ -293,7 +293,8 @@ func (mc *Chain) generateBlock(ctx context.Context, b *block.Block,
 
-	if config.DevConfiguration.IsBlockRewards {
+	if config.DevConfiguration.IsBlockRewards &&
+		b.Round%config.SmartContractConfig.GetInt64("smart_contracts.storagesc.block_reward.challenge_period") == 0 {
```
Instead of getting the setting from the viper config directly, we'd better put the `challenge period` setting in the scConfig and get it from there.

ssardana96 replied:
but @peterlimg we don't have access to storagesc.scConfig from here.

peterlimg replied:
Add method to the StateContextI to get the config, the config should be read from MPT, it could be different from the value read from the viper. We should always use the loaded config instead of reading it from viper directly.

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_list.go:
```diff
@@ -0,0 +1,162 @@
+//------------------------------------------------------------------------------
+
+type BlobberNode struct {
+	Id  string `json:"id"`
```
Id -> ID, to pass the lint, https://github.com/golang/go/wiki/CodeReviewComments#initialisms

Sriep replied:
It seems [contentious](https://stackoverflow.com/questions/15526107/acronyms-in-camelcase) but we should be consistent. Also, it's the first time I have seen [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments#go-code-review-comments) thanks for the reference.

ssardana96 replied:
fixed. this is helpful @peterlimg

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_reward.go:
```diff
@@ -0,0 +1,176 @@
+//------------------------------------------------------------------------------
+
+type BlobberRewardNode struct {
+	Id                string         `json:"id"`
```
Id -> ID

peterlimg commented on code/go/0chain.net/smartcontract/partitions/random.go:
```diff
@@ -205,6 +207,14 @@ func (rs *randomSelector) addPartition() PartitionItemList {
+		newPartition = &blobberItemList{
+			Key: rs.partitionKey(rs.NumPartitions),
+		}
+	} else if rs.ItemType == ItemBlobberReward {
```
Partition package should not be coupled with the concrete node types.

ssardana96 replied:
but @peterlimg partitions are of different for different itemtype
hence using it

peterlimg replied:
NVM, we can refactor this later. Anyway, I think each type of partition should has it's own RandomSelector instance, therefore, each RandomSelector can run related interface to generate its own type of partition item list instead of using the `if else..` here.

ssardana96 replied:
got it.
@peterlimg can you create an issue on the exact thought you have in mind. We can pick it up later

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -301,18 +349,28 @@ func (sng StorageNodeGeolocation) validate() error {
+	PublicKey         string                 `json:"-"`
 	// StakePoolSettings used initially to create and setup stake pool.
 	StakePoolSettings stakePoolSettings `json:"stake_pool_settings"`
+	RewardPartition   partitionLocation `json:"rewardPartition"`
```
Use underscore json tag

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -324,6 +372,56 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+					"cannot get ongoing partition: "+err.Error())
+			}
+
+			_, err = balances.InsertTrieNode(ACTIVE_PASSED_BLOBBERS_KEY, ongoingList)
```
If the active blobbers is a copy of the ongoingList and we can get it base on the  key calculated from start round, please do not save it to MPT again as a new 'active' list.  

So basically the `ongoing` and `active` blobber lists are logic concept, we should not put the the names into the MPT node key, and should avoid saving the duplicate data into MPT. 

What we should instead are having two functions:
func getOngoingPassedBlobbersList() {
   ...
   // the partition of current period
}

func getActivePasssedBlobbersList() {
  // get partition of previous period
}

There's no point to get the previous period partition and save it into a new partition that named `active`.

ssardana96 replied:
fixed @peterlimg 
can you check the new changes for this

peterlimg replied:
Thanks, I'll have a check

ssardana96 replied:
@peterlimg I have already made these changes long back.
if these are looking as expected can you mark this resolved?

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/blobber.go:
```diff
@@ -452,6 +485,51 @@ func (sc *StorageSmartContract) commitBlobberRead(t *transaction.Transaction,
+	if blobber.DRUpdateLastRound >= startRound {
+		blobber.DataReadLR += sizeRead
+	} else {
+		blobber.DataReadLR = sizeRead
```
If  the DataReadLR will be reset for each period, then we should save it in reward node. And we can remove the DRUpdateLastRound from blobber node as well.

ssardana96 replied:
@peterlimg we add rewardnode to partition only when challenge is passed.
What if read happens first, the rewardNode won't be in partition
but once challenge comes later on, then the reward node would be created and we would need this data

peterlimg replied:
Oh, yes, that makes scene. Blobber would have to keep the read data before it pass the challenge. 

One question: 
Do we have mechanism to ensure that all blobbers will have the chance to do verifyChallenge in a block reward period. So if a blobber could pass the challenge itself, but it just didn't get the chance, the code about seems would reset the `data read` of the blobber, should we reset it or accumulate the `data read` for next reward period? @Sriep @ssardana08

ssardana96 replied:
I believe we care only for what happened in the last period and hence reset.
@Sriep can comment more..

peterlimg commented on code/go/0chain.net/core/maths/maths.go:
```diff
@@ -0,0 +1,26 @@
+// A, B, alpha are constants
+// X is total data stored in blobber
+// R is data read by blobber
+func GetGamma(A, B, alpha, X, R float64) float64 {
```
What's the idea behind these Gama and Zeta, please link to document or paper properly to help us understand it, thanks.

Sriep replied:
Maybe you could take the pdf from slack, https://0chain.slack.com/archives/C02MT6RV6QP/p1644236441266629, and add it to `0chain/docs` then you can reference it directly.

peterlimg replied:
Thanks, I'll have a check.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -4,68 +4,103 @@ import (
+		// constants for zeta
+		I  = 1
+		K  = 1
+		mu = 1
```
If the above constants are used by functions defined in maths package, please define them in there too. Otherwise, users would have to define these in their own packages everywhere.

Sriep replied:
These six constants need to be added to smart contract settings read from `sc.yaml` Check slack thread. For six parameters it might be worth creating a separate object that can be passed to the maths package.

ssardana96 replied:
for now we're keeping these as it is, right @Sriep ?

peterlimg replied:
Maybe add a TODO comment or create an issue to track this

ssardana96 replied:
added a TODO and created an issue: https://github.com/0chain/0chain/issues/1035

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -4,68 +4,103 @@ import (
-	var stakeTotals []float64
-	var totalQStake float64
-	for _, blobber := range allBlobbers.Nodes {
+	random := rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
```
This random will make the block reward transaction got different results when it is executed on different miners. All miners should have the same random value so that the same blobbers will be selected for rewards. 

Please check how round random seed is generated and used to select the block generators.

Sriep replied:
https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/challenge.go#L570
Here the current transaction hash and previous block hash are used to generate the seed. You could turn this code from `generateChallenges` into a common function and use that.

peterlimg replied:
Yes, that should work as well and easy to go.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -84,3 +119,9 @@ func (ssc *StorageSmartContract) payBlobberBlockRewards(
 	return nil
 }
+
+func getBlockReward(br state.Balance, currentRound, brChangePeriod int64, brChangeRatio, blobberWeight float64) float64 {
```
The conf in line 114 is not changed, so we don't have to save into MPT again.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -145,6 +148,28 @@ func (sn *BlobberChallenge) addChallenge(challenge *StorageChallenge) bool {
+
+	for i := len(sn.Challenges) - 1; i >= 0; i++ {
+		if sn.Challenges[i].Created < fromTimestamp {
+			return
```
Return or continue?

ssardana96 replied:
not required function

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -145,6 +148,28 @@ func (sn *BlobberChallenge) addChallenge(challenge *StorageChallenge) bool {
+		}
+		success, failure := sn.Challenges[i].getStats()
+		pass, _, _ := sn.Challenges[i].isChallengePassed(success, failure, cct, now)
+		if pass == isSuccess {
```
Is it possible the `isSuccess` false? otherwise, we don't need this param I think.

ssardana96 replied:
this function is not needed

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -157,6 +182,29 @@ type StorageChallenge struct {
+	threshold = len(sc.Validators) / 2
+	pass = success > threshold ||
+		(success > failure && success+failure < threshold)
+	fresh = sc.Created+cct >= now
```
We should always require success > threshold

ssardana96 replied:
@peterlimg this was existing code, which I moved to a function as this was needed earlier.
reverting it back to older state

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -301,18 +349,28 @@ func (sng StorageNodeGeolocation) validate() error {
 	return nil
 }
 
+type partitionLocation struct {
```
I guess this is for blobber partition location only, if so, please do not use the common name (partition location) for it, rename it to properly such as `RewardPartitionLocation`.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -301,18 +349,28 @@ func (sng StorageNodeGeolocation) validate() error {
+	Capacity          int64                  `json:"capacity"`          // total blobber capacity
+	Used              int64                  `json:"used"`              // allocated capacity
+	BytesWritten      int64                  `json:"dataWritten"`       // in bytes
+	DataReadLR        float64                `json:"dataRead"`          // in GB
```
Rename to DataReadLastRound if this is needed, so as the json tag to:  data_read_last_round

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -301,18 +349,28 @@ func (sng StorageNodeGeolocation) validate() error {
+	Used              int64                  `json:"used"`              // allocated capacity
+	BytesWritten      int64                  `json:"dataWritten"`       // in bytes
+	DataReadLR        float64                `json:"dataRead"`          // in GB
+	DRUpdateLastRound int64                  `json:"dataReadLastRound"` // last round when data read was updated
```
Rename to DataReadUpdateLastRound if this is needed, json tag to: data_read_update_last_round

peterlimg reviewed (changes requested):
There are three main issues:
- The active and ongoing blobber reward lists, they should be logic concept, the name should not be insert into the MPT as part of the key. Also, we should not save the previous period data into MPT again as the `active` list.
- The random value generated in blobberBlockRewards is incorrect, it would lead to different transaction result get returned when it is executed on different miners/sharders.
- https://0chain.slack.com/archives/C020E8HA8TE/p1644062465226779?thread_ts=1643724926.281819&cid=C020E8HA8TE. The `partition` package should be a common package that we can use it easily to support partition, see the comments . Otherwise, we will have to do the things like the new added `blobber_list.go` and `blobber_reward.go`, basicaily most of the code there are duplicate work, we should avoid doing this. Anyway, this could be refactored later.

Sriep commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -145,6 +148,28 @@ func (sn *BlobberChallenge) addChallenge(challenge *StorageChallenge) bool {
 	return false
 }
 
+func (sn *BlobberChallenge) getSuccessCount(isSuccess bool,
```
This method does not seem to be used. If so, it is probably best to remove it.

ssardana96 replied:
removed

Sriep commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -157,6 +182,29 @@ type StorageChallenge struct {
+	return
+}
+
+func (sc *StorageChallenge) getStats() (success, fail int) {
```
`getStats` is only called from `getSuccessCount` which appears to be never used. So this code is unreachable. If so we should remove it.

ssardana96 replied:
removed

Sriep commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -157,6 +182,29 @@ type StorageChallenge struct {
 }
 
+func (sc *StorageChallenge) isChallengePassed(success, failure int,
+	challengeCompletionTime time.Duration, now common.Timestamp) (pass, fresh bool, threshold int) {
```
`isChallengePassed` can only be accessed from `verifyChallenge`. It does help to make the code more readable, however, the `BlobberChallege` is too big and will likely be removed as part of the concurrent changes for https://github.com/0chain/0chain/issues/741.  

So this code will be removed shortly. I think it would be best to not make unneeded changes to the `StorageChallenge` or `BlobberChallenge` object to reduce merging issues in the near future.

ssardana96 replied:
@Sriep this was intended changes made for previous version of BBR. will remove them

Sriep commented on code/go/0chain.net/smartcontract/storagesc/block_reward_test.go:
```diff
@@ -1,286 +1,124 @@
-				minerRatio:           2,
-				sharderRatio:         3,
+			name: "1_blobber",
+			params: params{
```
Once we have added the calculation parameters to sc.yaml settings, we should have a more realistic test case. With a full partition of blobbers, and add (gamma, a, b, i, k mu) to the `parmas` object. This could allow a user to experiment with different values.

ssardana96 replied:
to keep as constant for now

Sriep commented on docker.local/config/sc.yaml:
```diff
@@ -149,11 +149,13 @@ smart_contracts:
-      blobber_capacity_ratio: 10
-      blobber_usage_ratio: 40
+      blobber_ratio: 50
+      challenge_period: 30
```
Add calculation parameters (gamma, a, b, i, k, mu). Check slack thread https://0chain.slack.com/archives/C02MT6RV6QP/p1644840052762349?thread_ts=1644236441.266629&cid=C02MT6RV6QP for labels.

ssardana96 replied:
to be kept as it is for now

Sriep reviewed (changes requested):
As [verifyChallenge](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/challenge.go#L323) will be undergoing changes as part of the concurrent issue https://github.com/0chain/0chain/issues/741, it might be best to keep changes here minimal and focused on only what is needed for block rewards. This is to reduce any short term merging issues.

ssardana08 committed:
- some neats

ssardana08 committed:
- code refactor

ssardana08 committed:
- rewardPartitionLocation

ssardana08 committed:
- key names updated

ssardana08 committed:
- config shouldn't be re-saved in block reward

ssardana08 committed:
- code refactor
- making reward partition call more logical
- remove ongoing/active from key name

ssardana08 committed:
- trigger period added

ssardana08 committed:
- added method

ssardana08 committed:
- added method

peterlimg commented on code/go/0chain.net/miner/protocol_block_main.go:
```diff
@@ -293,7 +293,8 @@ func (mc *Chain) generateBlock(ctx context.Context, b *block.Block,
 
-	if config.DevConfiguration.IsBlockRewards {
+	if config.DevConfiguration.IsBlockRewards &&
+		b.Round%config.SmartContractConfig.GetInt64("smart_contracts.storagesc.block_reward.trigger_period") == 0 {
```
block_reward.period seems better, also get it from storagesc config variable instead of reading from viper.

ssardana96 replied:
I've changed it to block_reward.trigger_period @peterlimg 
on storagesc config we don't have access from here. Please suggest!

ssardana08 committed:
- block reward random see changed
- creating BlobberRewardKey with strings.Builder

Sriep commented on code/go/0chain.net/chaincore/chain/state/state_context.go:
```diff
@@ -58,6 +58,8 @@ type StateContextI interface {
 	GetEvents() []event.Event   // cannot use in smart contracts or REST endpoints
 	GetEventDB() *event.EventDb // do not use in smart contracts can use in REST endpoints
+	GetCurrentRewardRound(period int64) int64
+	GetPreviousRewardRound(period int64) int64
```
I cannot see why these two need to be in StateContextI. Cannot you just have a local function in stroagesc?
```go
func  GetCurrentRewardRound(period int64, balances cstate.StateContextI) int64 {
	extra := balances.GetBlock().Round % period
	return balances.GetBlock().Round - extra
}
```
Adding to StateContextI creates a maintenance overhead as it is used in so many places.

ssardana08 committed:
- removed GetCurrentRewardRound and GetPreviousRewardRound

ssardana08 committed:
- bbr unit tests fixed

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -4,68 +4,111 @@ import (
-	for _, blobber := range allBlobbers.Nodes {
+	hashString := encryption.Hash(balances.GetTransaction().Hash + balances.GetBlock().PrevHash)
+	var randomSeed uint64
+	randomSeed, err = strconv.ParseUint(hashString[0:16], 16, 64)
```
Use `ParseInt` so we don't  have to convert it to int64 below

peterlimg commented on code/go/0chain.net/smartcontract/utils/utils.go:
```diff
@@ -0,0 +1,11 @@
+package utils
```
These functions are blobber reward related, so it's better to put them in the package where blobber or reward structs are defined . `utils` is not a good package name actually, check the "Bad package names" section on https://go.dev/blog/package-names

ssardana96 replied:
done

ssardana08 committed:
- get current round/ get previous round checks added

ssardana08 committed:
- benchmark_setup fixed

ssardana08 committed:
- block reward updated

ssardana08 committed:
- check added

ssardana08 committed:
- using int64

ssardana08 committed:
- removed utils.go

ssardana08 committed:
- random fix

ssardana08 commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,149 @@ package storagesc
+
+	const (
+		// constants for gamma
+		alpha = 1
```
@Sriep for now it's okay to keep these constant?

ssardana08 committed:
- 0chain config changes

ssardana08 committed:
- struct property change

ssardana08 committed:
- block reward logging

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_list.go:
```diff
@@ -0,0 +1,143 @@
+	return nil
+}
+
+func (il *blobberItemList) add(it PartitionItem) {
```
We should check if the item does exist in the list before appending it, otherwise we could have duplicate items.

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,126 @@ package storagesc
+	}
+
+	for i, qsp := range stakePools {
+		reward := float64(conf.BlockReward.BlockReward) * (weight[i] / totalweight)
```
Check if  (weight[i] / totalWeight) > 0 and < 1

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,126 @@ package storagesc
+			} else {
+				ratio = 0
+			}
+			dReward := float64(conf.BlockReward.BlockReward) * weight[i] * ratio
```
explicitly check if ratio > 0 and < 1

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,106 @@ package storagesc
+			"cannot get all blobbers list: "+err.Error())
+	}
+
+	r := rand.New(rand.NewSource(time.Now().UTC().UnixNano()))
```
Any reason for changing back to use  time.Now() as random source?

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_list.go:
```diff
@@ -0,0 +1,162 @@
+
+func (il *blobberItemList) update(it PartitionItem) error {
+	var found bool
+	for i := range il.itemRange(0, il.length()) {
```
Why not just iterate through the item list, the itemRange will copy the list of item pointers, which seems like is not necessary

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/partitions/itemlist.go:
```diff
@@ -55,6 +56,21 @@ func (il *itemList) add(it PartitionItem) {
+	for i := range il.itemRange(0, il.length()) {
+		if il.Items[i].Name() == it.Name() {
+			found = true
+			il.Items[i] = StringItem{it.Name()}
```
we can set Changed=true and return nil here. And return the item not found error out the loop. No need to use the `found` variable, and should always return error early.

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/partitions/random.go:
```diff
@@ -320,3 +335,84 @@ func (rs *randomSelector) Decode(b []byte) error {
+			return err
+		}
+	} else {
+		return errors.New("partition to migrate to is already populated")
```
Check and returns error early to avoid this if else ... else mode

peterlimg commented on code/go/0chain.net/smartcontract/partitions/random.go:
```diff
@@ -320,3 +335,84 @@ func (rs *randomSelector) Decode(b []byte) error {
+				return err
+			}
+		} else {
+			return err
```
Check and returns error early to avoid this if else ... else mode

peterlimg commented on code/go/0chain.net/smartcontract/partitions/validator_list.go:
```diff
@@ -120,6 +121,24 @@ func (il *validatorItemList) add(it PartitionItem) {
+			found = true
+			il.Items[i] = ValidationNode{
+				Id:  it.Name(),
+				Url: it.Data(),
```
Return here with Changed=true, remove found variable

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,102 @@ package storagesc
+	}
+
+	// filter out blobbers with stake too low to qualify for rewards
+	var qualifyingBlobberIds []string
```
create the qulifyingBlobberIds slice with capacity of blobberPartition length

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,119 @@ package storagesc
+	}
+
+	for i, qsp := range stakePools {
+		reward := bbr * (weight[i] / totalWeight)
```
check explicitly to ensure the weight[i] / totalWeight is > 0 and < 1

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,119 @@ package storagesc
 }
+
+func getBlockReward(br state.Balance, currentRound, brChangePeriod int64, brChangeRatio, blobberWeight float64) float64 {
+	changeBalance := 1 - brChangeRatio
```
assert the changeBalance is > 0 and < 1

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,149 @@ package storagesc
+func GetCurrentRewardRound(currentRound, period int64) int64 {
+	if period > 0 {
+		extra := currentRound % period
+		if currentRound >= extra {
```
Is it possible currentRound < extra? it could happen only when it is < 0, but we should avoid it.

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/blobber.go:
```diff
@@ -461,6 +494,51 @@ func (sc *StorageSmartContract) commitBlobberRead(t *transaction.Transaction,
+	if blobber.LastRewardRoundOfDataRead >= rewardRound {
+		blobber.DataReadLastRewardRound += sizeRead
+	} else {
+		blobber.DataReadLastRewardRound = sizeRead
```
It's a little confuse to read the DataReadLastRewardRound, not sure if it is a round number or the data size from the name. So maybe we can change it to LastRewardDataReadSize for size. And change the LastRewardRoundOfDataRead to LastRewardDataReadRound for round.

ssardana96 replied:
yeah I know. I was myself thriving to get a better name. and hence added proper comments in model
LastRewardDataReadSize -  looks like the data size read in last reward, but as we're paying rewards in arrear, what we want to capture here is: what is the size read by blobber in last arrear reward round.

changed LastRewardDataReadRound

peterlimg commented on code/go/0chain.net/miner/protocol_block.go:
```diff
@@ -776,7 +776,9 @@ func (mc *Chain) generateBlock(ctx context.Context, b *block.Block,
 
-	if config.DevConfiguration.IsBlockRewards {
+	if config.DevConfiguration.IsBlockRewards &&
+		b.Round%config.SmartContractConfig.GetInt64("smart_contracts.storagesc.block_reward.trigger_period") == 0 {
```
Get config from MPT, anyway, should avoid read from file directly.

ssardana96 replied:
to be done later

ssardana08 committed:
- verify challenge logging added

ssardana08 committed:
- added block reward detailed checks and minor cleans

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -106,16 +113,19 @@ func (ssc *StorageSmartContract) blobberBlockRewards(
-		if err := qsp.DistributeRewards(reward, qualifyingBlobberIds[i], stakepool.Blobber, balances); err != nil {
-			return common.NewError("blobber_block_rewards_failed", "minting capacity reward"+err.Error())
+		weightRatio := weight[i] / totalWeight
+		if weightRatio > 0 && weightRatio <= 1 {
```
Invalid ratio should be reported or return as error if necessary.

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -106,16 +113,19 @@ func (ssc *StorageSmartContract) blobberBlockRewards(
-			return common.NewError("blobber_block_rewards_failed", "minting capacity reward"+err.Error())
+		weightRatio := weight[i] / totalWeight
+		if weightRatio > 0 && weightRatio <= 1 {
+			reward := bbr * (weight[i] / totalWeight)
```
Use the `weightRatio` instead of do it again with `weight[i] / totalWeight`

ssardana96 replied:
done

ssardana08 committed:
- paritionNode update function updated

ssardana08 committed:
- iterating through length of itemlist

ssardana08 committed:
- update method updated

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_list.go:
```diff
@@ -0,0 +1,159 @@
+	for i := range il.itemRange(0, il.length()) {
+		if il.Items[i].Name() == it.Name() {
+			var newItem BlobberNode
+			err := newItem.Decode(it.Encode())
```
We can do a type assertion and then do a copy to avoid unnecessary data encoding/decoding, such as:

bn, ok :=  (*it.BlobberNode)
if !ok{
  return err 
}

newItem :=  *bn
// add the newItem to list.

ssardana96 replied:
done

ssardana08 committed:
- added validation to check if the data already exists

ssardana08 committed:
- update methods updated with typecast

ssardana08 committed:
- weight ratio updates

ssardana08 committed:
- stakepool constant fixed

peterlimg commented on code/go/0chain.net/smartcontract/partitions/blobber_list.go:
```diff
@@ -0,0 +1,165 @@
+func (il *blobberItemList) add(it PartitionItem) {
+	for _, bi := range il.Items {
+		if bi.ID == it.Name() {
+			return
```
We'd better return error if the item already exist so that the caller could know what to do next. Imagine what could happen if we want to add a new item but the same name item does already exist in the list, then new item would not be inserted or updated to the exist one, then all following action against this new item will break. 

Same as other lists

ssardana96 replied:
done

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,172 @@ package storagesc
+			qsp.Reward += state.Balance(reward) // do we need to do this?
+		} else {
+			logging.Logger.Error("blobber_bloc_rewards - error in weight ratio",
+				zap.Any("stake pool", qsp))
```
We need to return error instead of continue for loop of reward calucation to avoid potential reward calculation issue.

ssardana96 replied:
done

ssardana08 committed:
- error return for weight ratio

ssardana08 committed:
- return error on add

ssardana08 committed:
- making qualifying blobber ids at init

ssardana08 committed:
- unit test fixed

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -2,10 +2,173 @@ package storagesc
+				return common.NewError("blobber_block_rewards_failed", "minting capacity reward"+err.Error())
+			}
+
+			qsp.Reward += state.Balance(reward) // do we need to do this?
```
I found the stake pool service reward has already been handled in the qsp.DistributeRewards, and it is the  serviceChargeRatio*reward, should we give all the reward to it again, this could be wrong.

Sriep replied:
The reward has already been paid by  `qsp.DistributeRewards`, so we should remove this line.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/challenge.go:
```diff
@@ -407,6 +474,64 @@ func (sc *StorageSmartContract) verifyChallenge(t *transaction.Transaction,
+
+			var dataRead float64 = 0
+			if blobber.LastRewardDataReadRound >= rewardRound {
+				dataRead = blobber.DataReadLastRewardRound
```
Hmm, in what situation the blobber.ReadPatition.StartRound != rewardRound but the blobber.LastRewardDataReadRound  >= rewardRound could happen?

Perhaps we need to make sure the blobber.ReadPartition.StartRound and the LastRewardDataReadRound are updated correctly.

ssardana96 replied:
when in this reward round period, data has been read by blobber, but verify_challenge has not been called yet for that blobber.

which means- in current reward round period, data has been accessed, but blobber has not yet been added to partition.

peterlimg commented on code/go/0chain.net/smartcontract/storagesc/models.go:
```diff
@@ -303,19 +317,30 @@ func (sng StorageNodeGeolocation) validate() error {
+	Terms                   Terms                  `json:"terms"`       // terms
+	Capacity                int64                  `json:"capacity"`    // total blobber capacity
+	Used                    int64                  `json:"used"`        // allocated capacity
+	BytesWritten            int64                  `json:"dataWritten"` // in bytes
```
json tag dataWritten -> data_written

ssardana08 committed:
- some neats
- distribute rewards corrected

ssardana08 committed:
fixed conflict

ssardana08 committed:
- msgpack issue fixed

ssardana08 committed:
- error fixed

ssardana08 committed:
- returning with logging error if partition is empty

ssardana08 committed:
- changes to rewardpartitionlocation

ssardana08 committed:
- test fixed

ssardana08 committed:
- added bbr cost

ssardana08 committed:
- logging added for debugging

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana08 committed:
- setconfig

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana08 committed:
- path fixed

Sriep commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -1,11 +1,180 @@
+			"cannot get smart contract configurations: "+err.Error())
+	}
+
+	if conf.BlockReward.BlockReward == 0 {
```
When I run the benchmark program, block rewards exited here. We need to setup the block reward settings in [SetMockConfig](https://github.com/0chain/0chain/blob/staging/code/go/0chain.net/smartcontract/storagesc/benchmark_setup.go#L642)

ssardana08 committed:
- mockConfig populated

ssardana08 committed:
- bbr benchmark added

ssardana08 committed:
- removed unused variable

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana08 committed:
- msgpack updated

Sriep commented on code/go/0chain.net/smartcontract/storagesc/block_reward.go:
```diff
@@ -1,11 +1,178 @@
+			return common.NewError("blobber_block_rewards_failed",
+				"can't decode blobber reward node: "+err.Error())
+		}
+		if sp, err = ssc.getStakePool(b.Name(), balances); err != nil {
```
My benchmark of block rewards gave ~20ms. How much do we reduce this if we get stakepools first? Something like
```go
wg sync.WaitGroup
c := make(chan error)
for i, b := range blobberPartitions {
  wg.add(1)
  go func {
    defer wg. Done
    if sp, err = ssc.getStakePool(b.Name(), balances); err != nil { c <- err }
    mu.Lock()
    qsp = append(qsp, sp)
    mu.UnLock()
  } ()
}
wg.Wait()
  for err := range c {
        if err != nil {
            return err
        }
  }
```

ssardana96 replied:
it brought down a bit by few ms.
for me it came down from 24 -> 19. are you benchmarking on some linux bare metal server. I guess there it'll be a bit better timing.
I have pushed changes in this branch

ssardana08 committed:
- get spool in goroutine

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana08 committed:
- fix order of sp

ssardana08 committed:
Merge branch 'bbr' of github.com:0chain/0chain into bbr

ssardana08 committed:
- lint fix

ssardana08 committed:
- lint fix

peterlimg committed:
Merge branch 'staging' into bbr

ssardana08 committed:
- change partition size for block rewards to 5

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana96 committed:
Merge branch 'staging' into bbr

ssardana96 committed:
Merge branch 'staging' into bbr

dabasov reviewed (approved):
lgtm
